packages/cli/src/models/compiler/Compiler.ts

```diff
@@ -47,11 +48,22 @@ export async function compile(
   const compileVersionOptions = compileVersion ? { version: compileVersion } : null;
 
   // Run typechain if requested
-  if (resolvedOptions.typechain && resolvedOptions.typechain.enabled) {
+  if (resolvedOptions.typechain?.enabled) {
+    Loggy.spin(__filename, 'compile', 'compile-typechain', 'Generating typechain artifacts...');
     const result = compileResult as ProjectCompileResult;
-    const filesList = result && result.artifacts ? result.artifacts.map(c => c.contractName).join(',') : '*';
-    const filesGlob = `${resolvedOptions.outputDir}/${filesList}.json`;
-    await typechain(filesGlob, resolvedOptions.typechain.outDir, resolvedOptions.typechain.target);
+    let filesGlob: string;
+    if (!result?.artifacts) {
```

Comment:
The line `const { compileWithTruffle, compileWithSolc } = exports;` cause to lose all the type information and then `const result = compileResult as ProjectCompileResult;` wrongly resets it because real type of `compileResult` is `void | ProjectCompileResult` which leads to `if (!result?.artifacts) {` not checking for `result` being `undefined`.
It can be fixed by using the following statements:
```
const withTruffle: typeof compileWithTruffle = exports.compileWithTruffle;
const withSolc:  typeof compileWithSolc = exports.compileWithSolc;
```
instead of `const { compileWithTruffle, compileWithSolc } = exports;` which do maintain type information.
Also you cant use optional chaining operator on `void` and this `const compilePromise = useTruffle ? compileWithTruffle() : compileWithSolc(resolvedOptions);` doesn't make much sense given `compileWithTruffle` returns `void`. I suggest to change return value to `undefined` or not to capture it return value at all.

Reply:
Good comments, thanks!!